presto-common/src/main/java/com/facebook/presto/common/block/MapBlockBuilder.java

```diff
@@ -0,0 +1,635 @@
+    private final BlockBuilderStatus blockBuilderStatus;
+
+    private int positionCount;
+    private int[] offsets;
+    private boolean[] mapIsNull;
+    private final BlockBuilder keyBlockBuilder;
+    private final BlockBuilder valueBlockBuilder;
+    private final HashTables hashTables;
+
+    private boolean currentEntryOpened;
+
+    public MapBlockBuilder(
+            Type keyType,
+            Type valueType,
+            MethodHandle keyBlockEquals,
+            MethodHandle keyBlockHashCode,
+            BlockBuilderStatus blockBuilderStatus,
+            int expectedEntries)
+    {
+        this(
+                keyBlockEquals,
+                keyBlockHashCode,
+                blockBuilderStatus,
+                keyType.createBlockBuilder(blockBuilderStatus, expectedEntries),
+                valueType.createBlockBuilder(blockBuilderStatus, expectedEntries),
+                expectedEntries,
+                null);
+    }
+
+    private MapBlockBuilder(
+            MethodHandle keyBlockEquals,
+            MethodHandle keyBlockHashCode,
+            @Nullable BlockBuilderStatus blockBuilderStatus,
+            BlockBuilder keyBlockBuilder,
+            BlockBuilder valueBlockBuilder,
+            int expectedEntries,
+            @Nullable int[] rawHashTables)
+    {
+        this.keyBlockEquals = requireNonNull(keyBlockEquals, "keyBlockEquals is null");
+        this.keyBlockHashCode = requireNonNull(keyBlockHashCode, "keyBlockHashCode is null");
+        this.blockBuilderStatus = blockBuilderStatus;
+
+        this.positionCount = 0;
+        this.offsets = new int[expectedEntries + 1];
+        this.mapIsNull = new boolean[expectedEntries];
+        this.keyBlockBuilder = requireNonNull(keyBlockBuilder, "keyBlockBuilder is null");
+        this.valueBlockBuilder = requireNonNull(valueBlockBuilder, "valueBlockBuilder is null");
+        this.hashTables = new HashTables(Optional.ofNullable(rawHashTables), 0);
+        this.logicalSizeInBytes = -1;
+    }
+
+    @Override
+    protected Block getRawKeyBlock()
+    {
+        return keyBlockBuilder;
+    }
+
+    @Override
+    protected Block getRawValueBlock()
+    {
+        return valueBlockBuilder;
+    }
+
+    @Override
+    protected HashTables getHashTables()
+    {
+        return hashTables;
+    }
+
+    @Override
+    protected int[] getOffsets()
+    {
+        return offsets;
+    }
+
+    @Override
+    public int getOffsetBase()
+    {
+        return 0;
+    }
+
+    @Override
+    protected boolean[] getMapIsNull()
+    {
+        return mapIsNull;
+    }
+
+    @Override
+    public int getPositionCount()
+    {
+        return positionCount;
+    }
+
+    @Override
+    public long getSizeInBytes()
+    {
+        return keyBlockBuilder.getSizeInBytes() + valueBlockBuilder.getSizeInBytes() +
+                (Integer.BYTES + Byte.BYTES) * (long) positionCount +
+                Integer.BYTES * HASH_MULTIPLIER * (long) keyBlockBuilder.getPositionCount();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of `Integer` multiplication is cast into a larger type. If you think the result of the multiplication could exceed the size of an `Integer`, consider casting the arguments of the multiplication to avoid arithmetic overflows. For more information, see [CWE-190](https://cwe.mitre.org/data/definitions/190.html).